helios-client/src/main/java/com/spotify/helios/client/HeliosClient.java

```diff
@@ -534,10 +542,62 @@ public Builder setEndpointSupplier(final Supplier<List<Endpoint>> endpointSuppli
     }
 
     public HeliosClient build() {
-      return new HeliosClient(user, new RetryingRequestDispatcher(
-          endpointSupplier.get(), user,
-          MoreExecutors.listeningDecorator(getExitingScheduledExecutorService(
-              (ScheduledThreadPoolExecutor) newScheduledThreadPool(4), 0, SECONDS))));
+      return new HeliosClient(user, createDispatcher());
+    }
+
+    private RequestDispatcher createDispatcher() {
+      final ScheduledExecutorService executor = MoreExecutors.getExitingScheduledExecutorService(
+          (ScheduledThreadPoolExecutor) newScheduledThreadPool(4), 0, SECONDS);
+
+      final ListeningScheduledExecutorService listeningExecutor =
+          MoreExecutors.listeningDecorator(executor);
+
+      final RequestDispatcher dispatcher =
+          new DefaultRequestDispatcher(createHttpConnector(), listeningExecutor);
+
+      return new RetryingRequestDispatcher(dispatcher,
+          listeningExecutor,
+          new SystemClock(),
+          5,
+          TimeUnit.SECONDS);
+    }
+
+    private HttpConnector createHttpConnector() {
+      // ssh identities (potentially) used in authentication
+      final List<Identity> identities = new ArrayList<>();
+
+      Optional<AgentProxy> agentProxyOpt = Optional.absent();
+
+      try (final AgentProxy agentProxy = AgentProxies.newInstance()) {
+        agentProxyOpt = Optional.of(agentProxy);
+        for (final Identity identity : agentProxy.list()) {
+          if (identity.getPublicKey().getAlgorithm().equals("RSA")) {
+            // only RSA keys will work with our TLS implementation
+            identities.add(identity);
+          }
+        }
+      } catch (Exception e) {
+        // the user likely doesn't have ssh-agent setup. This may not matter at all if the masters
+        // do not require authentication, so we delay reporting any sort of error to the user until
+        // the servers return 401 Unauthorized.
+        // This is also why we have the overly broad catch on Exception.
+        log.debug("Couldn't get identities from ssh-agent", e);
+      }
+
+      final EndpointIterator endpointIterator = EndpointIterator.of(endpointSupplier.get());
+
+      final DefaultHttpConnector connector = new DefaultHttpConnector(endpointIterator, 10000);
+
+      if (agentProxyOpt.isPresent()) {
```

Comment:
`agentProxyOpt` might be set even if there are no identities (e.g. `agentProxy.list()` threw an exception) but maybe that's intentional?